mgmtsystem_nonconformity/tests/test_workflows.py

```diff
@@ -0,0 +1,222 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    OpenERP, Open Source Management Solution
+#    This module copyright (C) 2015 - present Savoir-faire Linux
+#    (<http://www.savoirfairelinux.com>).
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from openerp.tests import common
+
+
+class BreakSavePoint(Exception):
+    pass
+
+
+class TestModelNonConformity(common.TransactionCase):
+    def setUp(self):
+        super(TestModelNonConformity, self).setUp()
+        self.partner_m = self.env['res.partner']
+        self.conf_m = self.env['mgmtsystem.nonconformity']
+
+        self.partner_obj = self.partner_m.search([])
+
+    def try_states(self, obj, *methods):
+        """
+        Call multiple methods to check if we can move from
+        one state to an other.
+        """
+        with self.cr.savepoint():
+            for method in methods:
+                with self.assertRaises(Exception):
+                    getattr(obj, method)()
+
+    def try_signal(self, obj, signal_name, state, rollback=False):
+        """
+        Try to trigger a signal.
+
+        Then rollback to keep working on the same object. As
+        if nothing happened.
+
+        The savepoint get rolledback when there is an error.
+        We use a custom error to catch when there is no error.
+        And refresh the model to get back the changes from the
+        database.
+        """
+        try:
+            sid = self.cr.savepoint()
+            with sid:
+                obj.signal_workflow(signal_name)
+                obj.refresh()
+                self.assertEqual(obj.state, state)
+
+                if rollback:
+                    raise BreakSavePoint()
+        except BreakSavePoint:
+            obj.refresh()
```

Comment:
Why is it necessary to have a custom error? Is it because of the context manager? Maybe you could define the Exception within this function in order to keep it encapsulated.